js/ui/data_grid/ui.data_grid.column_chooser.js

```diff
@@ -234,14 +234,33 @@ exports.ColumnChooserView = columnsView.ColumnsView.inherit({
         };
     },
 
+    _isListItemSelected: function(key) {
+        if(this._columnChooserList) {
+            var nodes = this._columnChooserList.getNodes(),
+                resultNode;
+
+            $.each(nodes, function(_, node) {
+                if(node.key === key) {
+                    resultNode = node;
+                    return false;
+                }
+            });
+
+            return resultNode && resultNode.selected;
+        }
+    },
+
     _columnOptionChanged: function(e) {
         var optionNames = e.optionNames,
             isSelectMode = this.option("columnChooser.mode") === "select";
 
         this.callBase(e);
 
-        if(isSelectMode && optionNames.showInColumnChooser) {
-            this.render(null, true);
+        if(isSelectMode) {
+            var visible = this._columnsController.columnOption(e.columnIndex, "visible");
+            if(optionNames.showInColumnChooser || (optionNames.visible && this._isListItemSelected(e.columnIndex) !== visible)) {
```

Comment:
I think that _isListItemSelected is not necessary. It contains problems with band columns and the code will be simpler if to use the following condition:

if(isSelectMode && (optionNames.showInColumnChooser || optionNames.visible)) {